packages/salesforcedx-vscode-apex/src/oas/esrHandler.ts

```diff
@@ -0,0 +1,447 @@
+  private overwrite: boolean = false;
+  private originalPath: string = '';
+  private newPath: string = '';
+
+  constructor() {}
+
+  public async initialize(
+    isESRDecomposed: boolean,
+    processedOasResult: ProcessorInputOutput,
+    fullPath: [string, string, boolean]
+  ) {
+    this.isESRDecomposed = isESRDecomposed;
+    this.processedOasResult = processedOasResult;
+    this.oasSpec = processedOasResult.openAPIDoc;
+    this.overwrite = fullPath[0] === fullPath[1];
+    this.originalPath = fullPath[0];
+    this.newPath = fullPath[1];
+  }
+
+  /**
+   * Generates the ESR metadata document.
+   * This method handles the process of reading existing ESR content,
+   * updating it with the new OpenAPI specification, and writing it back to the file system.
+   * It also manages the creation of the corresponding YAML file if ESR decomposition is enabled.
+   * Additionally, it gathers metrics, opens a diff editor if needed, and creates entries in the problems tab.
+   *
+   * @param isClass - Indicates if the generation is for a class.
+   * @param generationHrDuration - The duration of the generation process.
+   * @param promptGenerationOrchestrator - The orchestrator for prompt generation.
+   * @returns An object containing properties and measures for telemetry.
+   */
+  public async generateEsrMD(
+    isClass: boolean,
+    generationHrDuration: [number, number],
+    promptGenerationOrchestrator: PromptGenerationOrchestrator
+  ) {
+    const orgVersion = await (await WorkspaceContextUtil.getInstance().getConnection()).retrieveMaxApiVersion();
+    let existingContent;
+    let namedCredential;
+    if (fs.existsSync(this.newPath)) {
+      existingContent = fs.readFileSync(this.newPath, 'utf8');
+    }
+
+    //Step 1: Let user choose or enter a named credential
+    if (!this.isVersionGte(orgVersion, '63.0')) namedCredential = await this.showNamedCredentialsQuickPick();
+
+    //Step 2: Build the content of the ESR Xml file
+    const updatedContent = await this.buildESRXml(existingContent, namedCredential);
+
+    //Step 3: Write OpenAPI Document to File
+    try {
+      fs.writeFileSync(this.newPath, updatedContent);
+      await vscode.workspace.openTextDocument(this.newPath).then((newDocument: vscode.TextDocument) => {
+        void vscode.window.showTextDocument(newDocument);
+      });
+      if (this.isESRDecomposed) {
+        await vscode.workspace
+          .openTextDocument(this.replaceXmlToYaml(this.newPath))
+          .then((newDocument: vscode.TextDocument) => {
+            void vscode.window.showTextDocument(newDocument);
+          });
+      }
+    } catch (error) {
+      throw new Error(nls.localize('artifact_failed', error.message));
+    }
+
+    //Step 2: Gather metrics
+    const props = {
+      isClass: `${isClass}`,
+      overwrite: `${this.overwrite}`
+    };
+
+    const [errors, warnings, infos, hints, total] = summarizeDiagnostics(this.processedOasResult.errors);
+
+    const measures = {
```

Comment:
These measures are not part of the ESR, so they should be initialized else where